File: app/services/YeloReader.java
package services;

import java.net.URI;
import java.net.URISyntaxException;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import models.ImdbApiMovie;
import models.Movie;
import models.YeloTVGids;
import models.YeloTVGids.Broadcast;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import play.Logger;
import play.libs.WS;
import play.libs.WS.HttpResponse;
import play.libs.WS.WSRequest;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;

public class YeloReader {

	private static final String BASE_URI = "http://yelo.be/";

	public List<Movie> read(final Date date){
    	
    	List<Movie> movies = new ArrayList<Movie>();
    	movies.addAll(readDay(date));
    	movies.addAll(readChannel("jim", date));
    	movies.addAll(readChannel("ned1", date));
    	movies.addAll(readChannel("ned2", date));
    	movies.addAll(readChannel("ned3", date));
    	
    	for(Movie movie:movies){
    		if(!sameDay(movie.start, date)){
    			throw new RuntimeException("Movie returned for wrong day" + movie.start + ", should be " + date);
    		}
    	}
    	
    	return movies;
	}

	private List<Movie> readChannel(final String channel, final Date date) {
		DateFormat dayFormat = new SimpleDateFormat("dd-MM-yyyy");
		String day = dayFormat.format(date);
		String guideUri = BASE_URI + "zenders/"+channel+"?date="+day;
		Document doc = readUri(guideUri);
    	
    	List<Movie> movies = new ArrayList<Movie>();
    	
    	Elements links = doc.select("a[class][href~=(/film/).*");
    	for(Element link:links){
    		Movie movie = Movie.findByUrl(switchToHdWhereProssible(link.absUrl("href")));
    		if(movie == null){
    			movie = parseChannelPageMovie(link);
    			//movie.year = getYear(movie.url);
    			movie.save();
    		}
    		movies.add(movie);
    	}
		return movies;
	}
	
	private List<Movie> readDay(final Date date) {
		List<Movie> movies = new ArrayList<Movie>();
		movies.addAll(readDayPart(date, 1));
		movies.addAll(readDayPart(date, 2));
		return movies;
	}
	
	private List<Movie> readDayPart(final Date date, final long part) {
		DateFormat dayFormat = new SimpleDateFormat("dd-MM-yyyy");
		String day = dayFormat.format(date);
		
//		String guideUriAjax = BASE_URI + "tv-gids/groep/vlaams?date=" + day +"&part="+part + "&orientation=vertical";
//		Document ajaxDoc = readUri(guideUriAjax, "application/json, text/javascript, */*");
//		System.err.println(ajaxDoc);
		
		String guideUri = BASE_URI + "detail/tvgids?date="+day;
		JsonElement element = readUriJson(guideUri, null);
		//printJson(element);
		
		Gson gson = new Gson();
		YeloTVGids gids = gson.fromJson(element, YeloTVGids.class);
		for(Entry<String,Broadcast> entry:gids.result.meta.pvrbroadcasts.entrySet()){
			Broadcast broadcast = entry.getValue();
			if(broadcast.is_serie == 0){
				System.out.println(new Date(broadcast.start_time*1000) + "@" + broadcast.channel_webpvr_id + " " + broadcast.title);
			}
		}
		
    	
    	List<Movie> movies = new ArrayList<Movie>();
//    	Element tvgidsDiv = doc.select("div[class=tvgids-lijst box]").first();
//    	Elements links = tvgidsDiv.select("a[href~=(/film/).*");
//    	for(Element link:links){
//    		Movie movie = Movie.findByUrl(switchToHdWhereProssible(link.absUrl("href")));
//    		if(movie == null){
//    			movie = parseGuidePageMovie(link);
//    			movie.year = getYear(movie.url);
//    			movie.save();
//    		}
//    		movies.add(movie);
//    	}
		return movies;
	}

	private void printJson(JsonElement element) {
		Gson gson = new GsonBuilder().setPrettyPrinting().create();
		System.out.println(gson.toJson(element));
	}
	
//	Integer getYear(final String moviePageUrl){
//		Document doc = readUri(moviePageUrl);
////		<span class="filmdata-titel">Jaar: </span>
////		<span class="filmdata-content">2010</span>
//
//		Element yearTitleElement = doc.select("span[class=filmdata-titel]:contains(Jaar:)").first();
//		Element yearElement = yearTitleElement.nextElementSibling();
//		Integer year = null;
//		try{
//			year = Integer.valueOf(yearElement.text());
//		}catch(NumberFormatException ex){
//			Logger.warn("Could not parse year: [%s]", yearElement.text());
//		}
//		return year;
//	}

	
	private Movie parseGuidePageMovie(Element link){
    	DateFormat hourFormat = new SimpleDateFormat("HH:mm");
    	DateFormat dayFormat = new SimpleDateFormat("dd-MM-yyyy");
		Element em = link.siblingElements().select("em").first();
		
		Movie movie = new Movie();
		movie.title = link.text().trim();
		URI movieUri;
		try {
			movieUri = new URI(link.absUrl("href"));
		} catch (URISyntaxException e) {
			throw new RuntimeException(e);
		}
		Map<String,String> params = splitParameters(movieUri);
		movie.url = switchToHdWhereProssible(movieUri.toString());
		movie.channel = params.get("channel");
		try {
			movie.start = dayFormat.parse(params.get("date"));
		} catch (ParseException e) {
			throw new RuntimeException(e);
		}
		try {
			Date hour = hourFormat.parse(em.text());
			movie.start = TimeUtil.merge(movie.start, hour);
		} catch (ParseException e) {
			throw new RuntimeException(e);
		}
		return movie;
	}
	
	private Movie parseChannelPageMovie(Element link){
		DateFormat hourFormat = new SimpleDateFormat("HH:mm");
		DateFormat dayFormat = new SimpleDateFormat("dd-MM-yyyy");
		Element em = link.parent().siblingElements().select("span[class=datum]").first();
		
		Movie movie = new Movie();
		URI movieUri;
		try {
			movieUri = new URI(link.absUrl("href"));
		} catch (URISyntaxException e) {
			throw new RuntimeException(e);
		}
		movie.title = link.attr("title")
				.replace("FILM OP 2: ", "")
				.replace("Z@ppbios: ", "")
				.replace("Sunday Movie: ", "")
				.replace("Saturday Movie: ", "").trim();
		Map<String,String> params = splitParameters(movieUri);
		try {
			movie.start = dayFormat.parse(params.get("date"));
		} catch (ParseException e) {
			throw new RuntimeException(e);
		}
		movie.url = switchToHdWhereProssible(movieUri.toString());
		movie.channel = params.get("channel");
		try {
			Date hour = hourFormat.parse(em.text());
			movie.start = TimeUtil.merge(movie.start, hour);
		} catch (ParseException e) {
			throw new RuntimeException(e);
		}
		return movie;
	}
	

	private boolean sameDay(Date day1, Date day2) {
		Calendar day1Cal = Calendar.getInstance();
		day1Cal.setTime(day1);
		Calendar day2Cal = Calendar.getInstance();
		day2Cal.setTime(day2);
		return day1Cal.get(Calendar.YEAR) == day2Cal.get(Calendar.YEAR) &&
			day1Cal.get(Calendar.MONTH) == day2Cal.get(Calendar.MONTH) &&
			day1Cal.get(Calendar.DAY_OF_MONTH) == day2Cal.get(Calendar.DAY_OF_MONTH);
	}
	
	private Document readUri(String uri) {
		return this.readUri(uri, null);
	}
	
	private Document readUri(String uri, final String accept) {
		Logger.info(uri);
		WSRequest req = WS.url(uri);
		if(accept != null){
			req.setHeader("accept", accept);
		}
    	HttpResponse response = req.get();
    	String html = response.getString();
    	Document doc = Jsoup.parse(html, BASE_URI);
		return doc;
	}
	
	private JsonElement readUriJson(String uri, final String accept) {
		Logger.info(uri);
		WSRequest req = WS.url(uri);
		if(accept != null){
			req.setHeader("accept", accept);
		}
    	HttpResponse response = req.get();
    	return response.getJson();
	}
	
	private Map<String,String> splitParameters(URI uri){
		Map<String,String> result = new HashMap<String, String>();
		String query = uri.getQuery();
		String[] params = query.split("&");
		for(String param:params){
			String[] keyValue = param.split("=");
			result.put(keyValue[0], keyValue[1]);
		}
		return result;
	}
	
	private ImdbApiMovie readOrFetch(final Movie movie){
		ImdbApiMovie imdbMovie = ImdbApiMovie.findByTitle(movie.title);
		if(imdbMovie != null){
			movie.imdb = imdbMovie;
			movie.save();
		}
		return imdbMovie;
	}
	
	private String switchToHdWhereProssible(final String url){
		return url.replace("channel=een", "channel=een-hd")
				.replace("channel=ketnetcanvas", "channel=ketnetcanvas-hd")
				.replace("channel=vtm", "channel=vtm-hd");
	}
}

File: app/services/YeloReader.scala
package services

import org.joda.time.DateMidnight
import org.joda.time.format.DateTimeFormat
import org.joda.time.Interval
import org.joda.time.LocalTime
import play.api.libs.json.JsValue
import play.api.libs.json.JsObject
import play.api.Logger
import org.jsoup.Jsoup
import scala.collection.JavaConversions._
import play.api.libs.ws.WS
import java.text.ParseException
import org.joda.time.DateTime

import scala.concurrent.ExecutionContext.Implicits.global
import models.Channel


object YeloReader {
  
  private val logger = Logger("application.yelo")
  
  private val dateFormat = DateTimeFormat.forPattern("dd-MM-YYYY")
  
  def fetchDay(day: DateMidnight, channelFilter: List[String] = List()) = {
    val urls = urlsForDay(day)
    logger.info("Fetching " + urls)
    val first = WS.url(urls(0)).get().map { response =>
      parseDay(day, response.json).filter(result => channelFilter.isEmpty || channelFilter.map(_.toLowerCase).contains(Channel.unify(result.channel).toLowerCase))
    }
    val second = WS.url(urls(1)).get().map { response =>
      parseDay(day, response.json).filter(result => channelFilter.isEmpty || channelFilter.map(_.toLowerCase).contains(Channel.unify(result.channel).toLowerCase))
    }
    for( f <- first; s <- second ) yield ( f ++ s )
  }
  
  private def urlsForDay(day: DateMidnight) = {
    List(
      "http://yelo.be/detail/tvgids?date="+dateFormat.print(day),
      "http://yelo.be/detail/tvgids?date="+dateFormat.print(day) + "&group=anderstalig"
    )
  }
  
  private def parseDay(day: DateMidnight, body: JsValue): Seq[YeloEvent] = {

    val channelMap = parseChannels(body)

    parseEvents(body, channelMap)

  }


  def parseEvents(body: JsValue, channelMap: Map[String, String]): Seq[YeloReader.YeloEvent] = {
    val events = (body \ "Result" \ "Events").as[String]
    val doc = Jsoup.parse(events)

    val pvrbroadcasts = (body \ "Result" \ "Meta" \ "pvrbroadcasts").as[JsObject]

    //println("bc = " + pvrbroadcasts)

    pvrbroadcasts.fields.map {
      case (eventIdentification, eventObject) =>
        val eventId = eventIdentification.toInt
        val title = (eventObject \ "title").as[String]
        val channelWebpvrIid = (eventObject \ "channel_webpvr_id").as[String]
        val startTime = (eventObject \ "start_time").as[String].toLong * 1000
        val endTime = (eventObject \ "end_time").as[String].toLong * 1000

        val eventDiv = Option(doc.getElementsByAttributeValue("id", "js-event-" + eventId).first())
        eventDiv.map {
          div =>
            val url = "http://yelo.be" + div.getElementsByTag("a").attr("href")
            val channel = channelMap.get(channelWebpvrIid).getOrElse("UNKNOWN")
            YeloEvent(eventId, title, url, channel, new Interval(startTime, endTime))
        }.getOrElse {
          logger.warn("No div found for %s %s at %s".format(eventId, title, new DateTime(startTime)))
          YeloEvent(eventId, title, null, "UNKNOWN", new Interval(startTime, endTime))
        }

    }
  }

  private def parseChannels(body: JsValue) = {
    val channels = (body \ "Result" \ "Channels").as[String]
    val channelsDoc = Jsoup.parse(channels)

    val imgs = channelsDoc.getElementsByTag("img")
    val imglist = imgs.subList(0, imgs.size).toList
    val channelMap = imglist.map {
      img =>
        val alt = img.attr("alt")
        val src = img.attr("src")
        val file = src.substring(src.lastIndexOf("/") + 1)
        val code = file.substring(0, 5)
        (code, alt)
    }.toMap
    channelMap
  }

  def toHDUrl(url: String): String = {
    url.replace("vtm", "vtm-hd")
      .replace("canvas", "canvas-hd")
      .replace("een", "een-hd")
      .replace("vier", "vier-hd")
      .replace("vijf", "vijf-hd")
      .replace("2be", "2be-hd")
  }

  case class YeloEvent (
      id:Int, 
      name:String, 
      url: String, 
      channel: String,
      interval: Interval){
    
    def toDateTime() = interval.getStart
  }
}